src/picai_prep/dcm2mha.py

```diff
@@ -497,22 +455,140 @@ def convert(self):
+        self._image = None
+        self._metadata = None
+        self.dicom_slice_paths: Optional[List[str]] = None
+
+        self.series_reader = sitk.ImageSeriesReader()
+        try:
+            self._update_dicom_list()
+        except MissingDICOMFilesError:
+            self.dicom_slice_paths = None
+            if (self.path / "dicom.zip").exists():
+                self.path = self.path / "dicom.zip"
+            else:
+                raise MissingDICOMFilesError(f'No DICOM slices found in {self.path}')
+
+    @property
+    def image(self):
+        if self._image is None:
+            self._image = self._read_image()
+        return self._image
+
+    @property
+    def metadata(self):
+        if self._metadata is None:
+            self._metadata = self._read_metadata()
+        return self._metadata
+
+    def _read_image(self, path: Optional[PathLike] = None) -> sitk.Image:
+        if path is None:
+            path = self.path
+        path = Path(path)
+
+        if path.name == "dicom.zip":
+            return self._read_image_dicom_zip(path=path)
+        try:
+            return self._read_image_sitk(path=path)
+        except RuntimeError:
+            return self._read_image_pydicom(path=path)
+
+    def _update_dicom_list(self, path: Optional[PathLike] = None):
+        if path is None:
+            path = self.path
+
+        self.dicom_slice_paths = self.series_reader.GetGDCMSeriesFileNames(str(path))
+
+        # verify DICOM files are found
+        if len(self.dicom_slice_paths) == 0:
+            raise MissingDICOMFilesError(self.path)
+
+        if self.verify_dicom_filenames:
+            # verify DICOM filenames have increasing numbers, with no gaps
+            filenames = [os.path.basename(dcm) for dcm in self.dicom_slice_paths]
+            self._verify_dicom_filenames(filenames)
+
+    def _read_image_sitk(self, path: Optional[PathLike] = None) -> sitk.Image:
+        """
+        Read image using SimpleITK.
+
+        Parameters
+        ----------
+        path: PathLike
+            path to the folder containing the DICOM slices. The folder should contain the DICOM slices,
+            or a zip file named "dicom.zip" containing the DICOM slices.
+            default: self.path
+
+        Returns
+        -------
+        image: SimpleITK.Image
+        """
+        if path is not None:
+            self.path = path
+            self._update_dicom_list(path=path)
+
+        # read DICOM sequence
+        self.series_reader.SetFileNames(self.dicom_slice_paths)
+        image: sitk.Image = self.series_reader.Execute()
+
+        # read metadata from the last DICOM slice
+        reader = sitk.ImageFileReader()
+        reader.SetFileName(self.dicom_slice_paths[0])
+        reader.LoadPrivateTagsOn()
+        reader.ReadImageInformation()
+
+        # set metadata
+        metadata = {key: reader.GetMetaData(key).strip() for key in reader.GetMetaDataKeys()}
```

Comment:
`self._metadata` should have the named metadata as well, while we do not set that to the image. If we were to partially fill `self._metadata`, the named metadata is not added later. Filling in the `self._metadata` later with metadata from the image is super quick, so no issue there.

The code is different because of the named metadata. We can circumvent this difference by changing the way metadata is stored internally: always with the DICOM tag rather than the named version of it. We can still transform the user's named tags to DICOM tags to retain the current user interface.